### ISSUE TITLE
Use `fun rec` to begin a recursive lambda

### ISSUE THREAD

**Issue by daveyostcom** [fsharp/fslang-suggestions#1264]
**I propose we** allow the use of `fun rec` to begin a recursive lambda, like this.  
```fsharp
let fibonacci max =
  (0, 1) ||> fun rec fib n n' ->
    if n > max then []
    else n::(fib n' (n + n'))
```
A tuple function argument also works, if you use `|>` instead of `||>` (for two arguments) or `|||>` (for three arguments).
```fsharp
let fibonacci max =
  (0, 1) |> fun rec fib (n, n') ->
    if n > max then []
    else n::(fib (n', (n + n')))
```
When there are four or more input values, a tuple function argument must be used because F# currently lacks a pipe operator for that situation.

Despite the fact that the function has a name, which is necessary for recursion, the function is nevertheless a lambda; that is, the function is anonymous to surrounding code; the name can be used only inside the lambda.  Actually, it is possible today to write a recursive lambda (using a name).  See below.

Another example:
```fsharp
let takeWhileInclusive predicate inclusive (s : seq<_>) = 
  seq {
    use en = s.GetEnumerator()
    yield! fun rec loop() -> seq {
      if en.MoveNext() then
        if predicate en.Current then
          yield  en.Current
          yield! loop()
        elif inclusive then
          yield en.Current }
```

**The existing way of approaching this problem in F# is** to define a recursive function then immediately call it, typically only once:
```fsharp
let fibonacci max =
  let rec fib n n' =
    if n > max then []
    else n::(fib n' (n + n'))
  fib 0 1
```
There is also this option, which acts like a lambda, in that the name `fib` is not exposed, but it is less concise than the proposal:
```fsharp
let fibonacci max =
  (0, 1) ||> (let rec fib n n' =
                if n > max then []
                else n::(fib n' (n+ n'))
              in fib)
```

## Pros and Cons 

**The advantages of making this adjustment to F# are**
- allows any function to be a lambda, not just non-recursive functions
- aids chunking by presenting one chunk of code instead of two
- eliminates a line of code in most cases
- reads as a left-to-right data flow
- fits into pipeline style
- recognizable as a `for` loop use case because initial values are at the top
- understandable at sight by anyone who understands `fun` and `rec`

**The disadvantages of the old way**, where a `let` is followed by a call, are
- hinders chunking by presenting two chunks of code instead of one
- uses an extra line of code in most cases
- doesn’t read as a left-to-right data flow
- doesn’t fit with pipeline style
- obscures recognition of a `for` loop use case because initial values are at the bottom

**The disadvantages of making this adjustment to F# are**
- there is already a way to do the same thing

## Extra information

**Estimated cost (XS, S, M, L, XL, XXL):**  S

**Related suggestions:**
- [Make the `fun` keyword optional #168](https://github.com/fsharp/fslang-suggestions/issues/168)
- Nice to have: make `||>` work for a tuple of any arity

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by konst-sh** [comment]
I would guess that anonymous functions cannot be recursive for a reason, because they are anonymous and you can't call it by name (unlike in JS for example where you can assign a name to anonymous function and thus call it from inside). Your example, even if it be the correct expression is not recursive, it is just a function of 3 arguments with partially applied first and second ones which expects the last function and evaluates in a single pass.

**Comment by Martin521** [comment]
Well, I do think it would be doable. It is a 2 arg function, "fib" is the name.
And I like conciseness. And I have written such recursive functions and felt they are less concise than I wished.
But ... 
a) I don't think this rare case and the little you can gain does not warrant a language change whatsoever and 
b) The proposed change turns the `fun` keyword and lambda (= "anonymous") construct into a named one. This is confusing, certainly for newcomers.

**Comment by daveyostcom** [comment]
> (Fun fact: in terms of character count, the proposed alternatives are less concise than what we have today.)

Oh!  I hadn’t thought of that. Upon reflection, I think what feels concise to me about it is that the code simply does its thing, without the clutter of binding and then using a name at a distance.  Another way of saying this is that the code plugs into the left-to-right flow like a Lego block.

**Comment by cmeeren** [comment]
Not saying I'm in favor of this, just want to suggest that the recursive call could be done using `rec` instead of defining a name.

Using the OP example:

```f#
let fibonacci max =
  (0, 1) ||> fun rec n n' ->
    if n > max then []
    else n::(rec n' (n + n'))
```

**Comment by LyndonGingerich** [comment]
What, conceptually, is a lambda? I thought it was a one-time-use, unnamed function by definition. Recursion requires referencing a function multiple times, which requires some conceptual reference to that function; the only current system I know is to give it a name of some sort, thus making it no longer a conceptual lambda. (A lambda can be assigned a name for reuse, but then the caller is just going to think of it as a normal function even if it was defined as a lambda.)

In other words, instead of:
```fsharp
let fibonacci max =
  (0, 1) ||> fun rec fib n n' ->
    if n > max then []
    else n::(fib n' (n + n'))
```
Why not just define the function right at the call site, like this?
```fsharp
let fibonacci max =
  let rec fib n n' =
    if n > max then []
    else n::(fib n' (n + n'))
  (0, 1) ||> fib
```
Or, to avoid reuse elsewhere and ensure closeness to use position:
```fsharp
let fibonacci max =
  let rec fib n n' =
    if n > max then []
    else n::(fib n' (n + n'))
  in
  (0, 1) ||> fib
```

**Comment by daveyostcom** [comment]
> What, conceptually, is a lambda? I thought it was a one-time-use, unnamed function by definition.

So did I, and I think so did many of us.

Thanks, you inspired some improvements to the Pros and Cons section.

**Comment by daveyostcom** [comment]
> Not saying I'm in favor of this, just want to suggest that the recursive call could be done using the existing `rec` keyword instead of defining a name.

That or some new keyword.  It’s a bit jarring to overload a keyword that way, I think, whereas I see no problem with simply allowing the user pick a name, as is natural for non-lambda recursive functions.  IMO most will get used to the idea of a recursive lambda that has a name for internal use, and many will welcome it with 🤦.

**Comment by Martin521** [comment]
> ... or some new keyword

The `as` keyword might be more appropriate
```
let fibonacci max =
  (0, 1) |> fun (n, n') as fib ->
    if n > max then []
    else n::(fib (n', (n + n')))
```
Ceterum censeo ... see my comment above

**Comment by LyndonGingerich** [comment]
We could also do the C# thing and define a keyword like `this` to refer to the context in which it is used. But I think that concomitant chaos would outweigh any good of the change.

It sounds to me as though the real issue is the ability to use pipeline-style function application everywhere, or at least in recursion. Is this a good idea, though? Pipeline style is a useful tool with one argument, and it works with, say, DU case constructors with tuples, but with normal functions I find it confusing with more than one argument.

If what you really want is a non-mutating for-loop with initial state at the top, is there a way to write a computation expression for that?

**Comment by daveyostcom** [comment]
> ```fsharp
> (a, b)
> ||> (let rec f a b = f a b in f)
> ```
> 
> While clunky, the current syntax technically does so as well.

Interesting!  I did not know that.  So we can write write a recursive lambda after all:
```fsharp
let fibonacci max =
  (0, 1) ||> (let rec fib n n' =
                if n > max then []
                else n::(fib n' (n+ n'))
              in fib)
```
[Verbose syntax](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/verbose-syntax) is kind of off the beaten track.

**Comment by Thorium** [comment]
Do you remember this? https://web.archive.org/web/20120330034902/http://blogs.msdn.com//b/madst/archive/2007/05/11/recursive-lambda-expressions.aspx

Meanwhile nice for nerdy hobby, I'd like to avoid using y-combinator and fixed point in production.
